### PR TITLE
Use kernel-side route filtering for large performance boost

### LINKED ISSUE #5255: Long dataplane apply times in route table apply
## Expected Behavior

We've been investigating some high Dataplane Apply times to see what's going on.

Our environment have an average of 600 PODs per node. We've being trying to improve performance and make this possible, even it not being recommended ;)

The Dataplane apply time takes an average of 0.5s, but in some very specific times it takes more than 15s.

Making some digging and brakepointing into the code, we've faced that while applying the dataplane the routine that takes more time is [this](https://github.com/projectcalico/felix/blob/master/dataplane/linux/int_dataplane.go#L1124)

When digging more into this function, we've faced that the [following](https://github.com/projectcalico/felix/blob/master/routetable/route_table.go#L318-L367) part of the code, WHEN CALLED takes that huge amount of time.

What we mean by WHEN? Because this seems an interaction. Sometimes there's no Item to interact and then this piece of code is not called.

But when it's called, the whole array of the iteration calls the ``r.syncRoutesForLink(ifaceName)`` (and all of them with success).

While each call per interface of syncRoutesForLink takes an average of 26ms and a maximum of 113ms, the sum of the times for 400 interfaces is something like 15000ms (15s!).

The question/suggestion of improvement here is:

1) Couldn't those syncRoutesForLink be not serial, but instead run concurrently to drop those requests from 15s to a maximum of 115ms?

2) I couldn't figure out why all interfaces got a syncRoutesForLink instead of only the changed interface (I may be missing something, sorry :( )

It seems, btw that this routes WG does not block IPTables (https://github.com/projectcalico/felix/blob/release-v3.6/dataplane/linux/int_dataplane.go#L936,L949) or anything like that, so it doesn't impact into the rest of dataplane update. 

So may this represent a performance issue or this can be treated just as an improvement?

## Your Environment
* And/Or, `calico/node` container version (if running Felix in a container): 3.6.4
* Which orchestrator are you using (e.g. Kubernetes, OpenStack, Docker, Mesos): Kubernetes
* Orchestrator version (if used): 1.13.4
* Etcd version (if using etcd datastore driver): 3.3
* Operating System and version: CoreOS 2079.6.0

### PR DESCRIPTION
**Pull request by fasaxc**
## Description

Use kernel-side route filtering in kernels v4.10+ when listing routes.  This gives a dramatic performance boost when doing route resyncs.  Previously, the netlink library would, under the covers, load _all_ routes, resulting in really poor performance when we were expecting to only list 1-2 routes.

- Rev netlink library to version supporting the feature.
- Do feature detection to detect kernel v4.10+ (lots of dull plumbing to get the feature detector to all of our routing tables).
- Enable "strict" mode on the route tables' netlink handle.  This activates the feature by telling the kernel to check the extra kernel-mode filtering attributes.
- Handle different errors returned by the kernel-mode filtering logic.

## Related issues/PRs

Should help with #7290.
Should help with #7085.
Should help with #7029.
Fixes #5255

## Todos

- [x] Tests
- [x] Documentation
- [x] Release note

## Release Note

```release-note
Performance: use kernel-side route filtering when listing routes.  Dramatically reduces CPU usage (and garbage collection) on systems with many interfaces.
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

**Review by electricjesus — APPROVED** [review]
LGTM

File: felix/dataplane/linux/int_dataplane.go
Comment (nelljerram) [reply felix/dataplane/linux/int_dataplane.go:483]: After reviewing the whole change, I guess it was probably just that I hadn't made the right change widely enough - which you now **have** done.

**Review by nelljerram — APPROVED** [review]
A few nits and queries, but happy for you to decide them all.

File: felix/routetable/route_table.go
Comment (fasaxc) [reply felix/routetable/route_table.go:315]: That was my thought.  Avoid accidentally having two code paths now.

File: felix/routetable/route_table.go
Comment (fasaxc) [reply felix/routetable/route_table.go:534]: I think it's so old that it's actually hard to find; I'm inclined to fix if and when someone hits an issue (but the groundwork is laid at least).

File: felix/dataplane/linux/int_dataplane.go
Comment (fasaxc) [reply felix/dataplane/linux/int_dataplane.go:483]: Practically nothing.  I think it may have an impact if your kernel is compiled without multiple routing tables but that's certainly not something we support.

When the kernel sees 0 it defaults it to main anyway.

It does have a real impact on the kernel-side filtering logic though.  0 is the wildcard so the kernel returns all routes from all tables.  Then netlink filters it down with special-case logic in userspace.

File: felix/iptables/table.go
Comment (fasaxc) [reply felix/iptables/table.go:317]: I've flipped the naming so that the interface has the suffix. I think making concrete types private is a bit of an anti-pattern in Go; you're intended to define your interfaces locally for mocking but it got quite messy with the dozens of places I needed this interface.

File: felix/wireguard/wireguard.go
Comment (fasaxc) [reply felix/wireguard/wireguard.go:30]: grr.  I keep meaning to write a tool to make these disappear.  Very annoying that my editor and go-fmt interact this way

File: felix/iptables/table_test.go
Comment (fasaxc) [reply felix/iptables/table_test.go]: Ah yes, see what you mean; seems like the tests in the iptables package also cover the real feature detector so this is a bit fiddly to unpick.

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and leave the branch after I've merged it.

**Comment by marvin-tigera** [comment]
Removing "merge-when-ready" label due to new commits

**Pull request merged**
